featureflags.go

```diff
@@ -142,43 +202,382 @@ func (poller *FeatureFlagsPoller) IsFeatureEnabled(key string, distinctId string
-		if key == storedFlag.Key {
-			featureFlag = storedFlag
-			break
+	return true, nil
+}
+
+func matchProperty(property Property, properties Properties) (bool, error) {
+	key := property.Key
+	operator := property.Operator
+	value := property.Value
+	if _, ok := properties[key]; !ok {
+		return false, &InconclusiveMatchError{"Can't match properties without a given property value"}
+	}
+
+	if operator == "is_not_set" {
+		return false, &InconclusiveMatchError{"Can't match properties with operator is_not_set"}
+	}
+
+	override_value, _ := properties[key]
+
+	if operator == "exact" {
+		switch t := value.(type) {
+		case []interface{}:
+			return contains(t, override_value), nil
+		default:
+			return value == override_value, nil
+		}
+	}
+
+	if operator == "is_not" {
+		switch t := value.(type) {
+		case []interface{}:
+			return !contains(t, override_value), nil
+		default:
+			return value != override_value, nil
+		}
+	}
+
+	if operator == "is_set" {
+		return true, nil
+	}
+
+	if operator == "icontains" {
+		return strings.Contains(strings.ToLower(fmt.Sprintf("%v", override_value)), strings.ToLower(fmt.Sprintf("%v", value))), nil
+	}
+
+	if operator == "not_icontains" {
+		return !strings.Contains(strings.ToLower(fmt.Sprintf("%v", override_value)), strings.ToLower(fmt.Sprintf("%v", value))), nil
+	}
+
+	if operator == "regex" {
+
+		r, err := regexp.Compile(fmt.Sprintf("%v", value))
+
+		// invalid regex
+		if err != nil {
+			return false, nil
+		}
+
+		match := r.MatchString(fmt.Sprintf("%v", override_value))
+
+		if match {
+			return true, nil
+		} else {
+			return false, nil
 		}
 	}
 
-	if featureFlag.Key == "" {
-		return defaultResult, nil
+	if operator == "not_regex" {
+		var r *regexp.Regexp
+		var err error
+
+		if valueString, ok := value.(string); ok {
+			r, err = regexp.Compile(valueString)
+		} else if valueInt, ok := value.(int); ok {
+			valueString = strconv.Itoa(valueInt)
+			r, err = regexp.Compile(valueString)
+		} else {
+			errMessage := "Regex expression not allowed"
+			return false, errors.New(errMessage)
+		}
+
+		// invalid regex
+		if err != nil {
+			return false, nil
+		}
+
+		var match bool
+		if valueString, ok := override_value.(string); ok {
+			match = r.MatchString(valueString)
+		} else if valueInt, ok := override_value.(int); ok {
+			valueString = strconv.Itoa(valueInt)
+			match = r.MatchString(valueString)
+		} else {
+			errMessage := "Value type not supported"
+			return false, errors.New(errMessage)
+		}
+
+		if !match {
+			return true, nil
+		} else {
+			return false, nil
+		}
 	}
 
-	return poller.getFeatureFlagVariant(featureFlag, key, distinctId)
+	if operator == "gt" {
+		valueOrderable, overrideValueOrderable, err := validateOrderable(value, override_value)
+		if err != nil {
+			return false, err
+		}
+
+		return overrideValueOrderable > valueOrderable, nil
+	}
+
+	if operator == "lt" {
+		valueOrderable, overrideValueOrderable, err := validateOrderable(value, override_value)
```

Comment:
i think we can probably consolidate these two lines higher up

Reply:
What do you mean?

Reply:
tbh these are just nits but given how overwhelming most of the local feature flags code is the cleaner the code the easier it is to digest but also it's not a high priority since the go library isn't used heavily so 🤷‍♀️